Lib/pdb.py

```diff
@@ -1961,6 +1982,23 @@ def do_debug(self, arg):
 
     complete_debug = _complete_expression
 
+    def do_attach(self, process):
+        """attach process
+
+        Attach to process, which can be a subprocess.Popen,
+        multiprocessing.Process or a pid.
+        """
+        pid = self._get_pid_from_process(process)
+
+        if pid is not None:
+            self.message(f"Attaching to process {pid}")
+            try:
+                attach(pid)
```

Comment:
There are some assumptions in `attach` that the `_PdbClient` is being run from the main thread - the entire signal handling approach relies upon it, at the very least - I think it'll just wind up failing with an exception if run from a non-main thread, at the point where it tries to install the signal handler.

I think we should check if we're in the main thread explicitly and `self.error()` if not.

Or, thinking outside of the box, we could spawn a new process to do the attaching from its main thread, and return control back to the parent process after it finishes.

Reply:
The great thing about attaching directly from the process is that there's to automatic parent-child relation. So even with the tracing restriction, you can still attach to your child processes - spawning a new process won't let you do that.

Reply:
That's true, and that's a nice advantage - but certainly not one big enough to live with Ctrl+C not working! And you only get that advantage when starting off in non-remote PDB - if you started off in remote PDB, the child process won't be a child of the client (which means it also won't work when you `attach` to a child process and then `attach` again to a grandchild process).

The most reasonable choice might just be to make it so that the `pdb.attach` module level function raises an exception if it's called from any thread but the main thread. Then the `attach` PDB command will work from any thread when running remote PDB, and from the main thread when running normal PDB, but will fail when when called on a non-main thread in non-remote PDB.

Reply:
> (which means it also won't work when you attach to a child process and then attach again to a grandchild process).

Is that true? I thought the trace_scope applies to descendants, not only direct children. So grandchildren should work?

Reply:
> The most reasonable choice might just be to make it so that the pdb.attach module level function raises an exception if it's called from any thread but the main thread.

That seems a reasonable option. It won't work when `_PdbClient` is not in the main thread. We can explore other options in the future, but for now, making what works work seems to be the way to go.

Reply:
> > (which means it also won't work when you attach to a child process and then attach again to a grandchild process).

> Is that true? I thought the trace_scope applies to descendants, not only direct children. So grandchildren should work?

Ah, yes, you're right.